packages/@lwc/ssr-compiler/src/transmogrify.ts

```diff
@@ -0,0 +1,183 @@
+        // - fallbackTmpl vs fallbackTmplNoYield
+        //
+        // If this becomes too burdensome to maintain, we can officially deprecate the generator-based approach
+        // and switch the @lwc/ssr-runtime implementation wholesale over to the no-generator paradigm.
+
+        const { node } = path;
+        if (!node || node.imported.type !== 'Identifier') {
+            throw new Error(
+                'Implementation error: unexpected missing identifier in import specifier'
+            );
+        }
+
+        if (
+            path.parent?.type !== 'ImportDeclaration' ||
+            path.parent.source.value !== '@lwc/ssr-runtime'
+        ) {
+            return;
+        }
+
+        if (node.imported.name === 'fallbackTmpl') {
+            node.imported.name = 'fallbackTmplNoYield';
+        } else if (node.imported.name === 'renderAttrs') {
+            node.imported.name = 'renderAttrsNoYield';
+        }
+    },
+};
+
+/**
+ * Transforms async-generator code into either the async or synchronous alternatives that are
+ * ~semantically equivalent. For example, this template:
+ *
+ *   <template>
+ *       <div>foobar</div>
+ *       <x-child></x-child>
+ *   </template>
+ *
+ * Is compiled into the following JavaScript, intended for execution during SSR & stripped down
+ * for the purposes of this example:
+ *
+ *   async function* tmpl(props, attrs, slottedContent, Cmp, instance) {
+ *       yield '<div>foobar</div>';
+ *       const childProps = {};
+ *       const childAttrs = {};
+ *       yield* generateChildMarkup("x-child", childProps, childAttrs, childSlottedContentGenerator);
+ *   }
+ *
+ * When transmogrified in async-mode, the above generated template function becomes the following:
+ *
+ *   async function tmpl($$emit, props, attrs, slottedContent, Cmp, instance) {
+ *       $$emit('<div>foobar</div>');
+ *       const childProps = {};
+ *       const childAttrs = {};
+ *       await generateChildMarkup($$emit, "x-child", childProps, childAttrs, childSlottedContentGenerator);
+ *   }
+ *
+ * When transmogrified in sync-mode, the template function becomes the following:
+ *
+ *   function tmpl($$emit, props, attrs, slottedContent, Cmp, instance) {
+ *       $$emit('<div>foobar</div>');
+ *       const childProps = {};
+ *       const childAttrs = {};
+ *       generateChildMarkup($$emit, "x-child", childProps, childAttrs, childSlottedContentGenerator);
+ *   }
+ *
+ * There are tradeoffs for each of these modes. Notably, the async-yield variety is the easiest to transform
+ * into either of the other varieties and, for that reason, is the variety that is "authored" by the SSR compiler.
+ */
+export function transmogrify(
+    compiledComponentAst: EsProgram,
+    mode: TransmogrificationMode = 'sync'
+): EsProgram {
+    const state: TransmogrificationState = {
+        mode,
+    };
+
+    return produce(compiledComponentAst, (astDraft) => traverse(astDraft, visitors, state));
```

Comment:
@wjhsf FYI we should use this instead of `structuredClone`: https://immerjs.github.io/immer/produce